tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp

```diff
@@ -308,12 +308,20 @@ class ASTBuildOperation
   /// can be determined at construction time of the \c ASTBuildOperation.
   const std::vector<FileContent> FileContents;
 
+  /// Guards \c DependencyStamps. This prevents reading from \c DependencyStamps
+  /// while it is being modified. It does not provide any ordering gurantees
+  /// that \c DependencyStamps have been computed in \c buildASTUnit before they
+  /// are accessed in \c matchesSourceState but that's fine (see comment on
+  /// \c DependencyStamps).
+  llvm::sys::Mutex DependencyStampsMtx;
+
   /// \c DependencyStamps contains the stamps of all module depenecies needed
   /// for the AST build. These stamps are only known after the AST is built.
   /// Before the AST has been built, we thus assume that all dependency stamps
   /// match. This seems to be a reasonable assumption since the dependencies
   /// shouldn't change (much) in the time between an \c ASTBuildOperation is
   /// created and until it produced an AST.
+  /// Must only be accessed if \c DependencyStampsMtx has been claimed.
   SmallVector<std::pair<std::string, BufferStamp>, 8> DependencyStamps = {};
```

Comment:
There's quite a bit of shared state in this class in general. Is this the only other one that needs the lock? Should we just merge this one and `ConsumersAndResultMtx` and make it a `SharedStateMtx` or something?